### PR TITLE
Update protobuf to 3.0.0 (GA) and the gRPC plugin to -pre1.2

### PR DESCRIPTION
**Pull request by jcanizales**
Note that the podspecs were already pushed this morning (and the zipped plugin already updated) - this just makes our repo match reality.

This was done to make some sample app (@makdharma knows about it) compile - the version of `protoc` we were using before didn't produce good code for `one_of`'s when integrating as dynamic frameworks.

### CODE REVIEW

**Comment by muxi** [comment]
[gRPC-ProtoRPC.podspec](https://github.com/grpc/grpc/blob/master/gRPC-ProtoRPC.podspec#L64) may need to be updated for beta-4

**Comment by jcanizales** [comment]
It's also blocked by this anyway https://github.com/grpc/grpc/issues/7479 so we're most probably going to have to use something newer than `-beta4` for GA.

File: third_party/protobuf [hunk 1]
```diff
@@ -1 +1 @@
-Subproject commit bdeb215cab2985195325fcd5e70c3fa751f46e0f
+Subproject commit 4b36284bdfdc77896d7c4cb9645433cf86efb069
```
Comment (jtattermusch) [inline third_party/protobuf]: it looks like this commit does not exist at all in google/protobuf repository.

v3.0.0-beta-4 seems to be commit 56855f6.
https://github.com/google/protobuf/releases

Comment (jtattermusch) [inline third_party/protobuf]: Correction: it is an existing commit, but it's not v3.0.0-beta4

**Comment by jcanizales** [comment]
Ok, now that Protobuf 3 is GA, I'm going to repurpose this PR to advance to that :)

**Comment by jcanizales** [comment]
Please test this.

**Comment by jcanizales** [comment]
Hey @y-zeng, what do I have to do to satisfy [this test you wrote](https://github.com/grpc/grpc/blob/v1.0.x/tools/distrib/check_generated_pb_files.sh) after updating the version of protobuf we use? It's failing on Jenkins.

**Comment by pgrosu** [comment]
@jcanizales It's also failing since [lcov is not properly installed](https://grpc-testing.appspot.com/job/gRPC_pull_requests/config=gcov,language=php7,platform=linux/11376/console), which is triggered when running the [`tools/run_tests/build_php.sh`](https://github.com/grpc/grpc/blob/v1.0.x/tools/run_tests/build_php.sh) script:

```
configure: error: To enable code coverage reporting you must have one of the following LCOV versions installed: 1.5 1.6 1.7 1.9 1.10 1.11
Command exited with non-zero status 1
real 1.50
user 0.82
sys 0.15

FAILED: tools/run_tests/build_php.sh [ret=1, pid=4827]
```

Hope it helps,
~p

**Comment by jcanizales** [comment]
@nicolasnoble: Can you check out what's up with lcov in the Jenkins runner?

**Comment by y-zeng** [comment]
Hi @jcanizales, you can run tools/codegen/extensions/gen_reflection_proto.sh to update the generated files.

File: src/objective-c/!ProtoCompiler-gRPCPlugin.podspec [hunk 2]
```diff
@@ -37,7 +37,7 @@ Pod::Spec.new do |s|
   # before them.
   s.name     = '!ProtoCompiler-gRPCPlugin'
   v = '1.0.0-pre1'
-  s.version  = v
+  s.version  = "#{v}.2" # .2 to depend on protoc 3.0.0
```
Comment (muxi) [inline src/objective-c/!ProtoCompiler-gRPCPlugin.podspec]: Why don't we just use v = '1.0.0-pre1.2' here :)

Comment (jcanizales) [inline src/objective-c/!ProtoCompiler-gRPCPlugin.podspec]: It should usually be `"#{v}"` - I started adding `.x` at the end just to be able and push updated versions to the Cocoapods repo. The intention is to get rid of the `.x` as soon as we have another pre-release, or at GA.

File: gRPC-ProtoRPC.podspec [hunk 3]
```diff
@@ -61,7 +61,7 @@ Pod::Spec.new do |s|
 
   s.dependency 'gRPC', version
   s.dependency 'gRPC-RxLibrary', version
-  s.dependency 'Protobuf', '~> 3.0.0-beta-3.1'
+  s.dependency 'Protobuf', '~> 3.0'
```
Comment (muxi) [inline gRPC-ProtoRPC.podspec]: The template for this file seems no longer exists. Update the comments at the beginning of this file?

**Comment by jcanizales** [comment]
Hey @jtattermusch, what does [this C# test failure](https://grpc-testing.appspot.com/job/gRPC_pull_requests/11764/config=opt,language=csharp,platform=windows/testReport/(root)/tests/csharp_Grpc_Core_Tests_SanityTest/) mean? If it's not a flake, how can it be fixed?

File: gRPC-ProtoRPC.podspec (hunk 3 above)
Comment (jcanizales) [inline gRPC-ProtoRPC.podspec]: Good catch; done.

**Comment by murgatroid99** [comment]
@apolcyn Can you take a look at that C# failure?

**Comment by pgrosu** [comment]
@jcanizales: Try the following update to [line 111 in SanityTest.cs](https://github.com/grpc/grpc/blob/f14effb6e421601639ef9a293f3f9b55a51e87ed/src/csharp/Grpc.Core.Tests/SanityTest.cs#L65):

```C#
Assert.AreEqual(discoveredTests.ToString().Replace('\r',''), testsFromFile.ToString());
```

or this more complete one:

```C#
Assert.AreEqual(discoveredTests.ToString().Replace('\r','').Replace('\n',''), testsFromFile.ToString().Replace('\r','').Replace('\n',''));
```

Based on the error the difference seems to be just that the `\r` (carriage return) is included:

```
1) Failed : Grpc.Core.Tests.SanityTest.TestsJsonUpToDate
  Expected string length 2087 but was 2036. Strings differ at index 1.
  Expected: "{\r\n  "Grpc.Core.Tests": [\r\n    "Grpc.Core.Internal.Tests.Asyn..."
  But was:  "{\n  "Grpc.Core.Tests": [\n    "Grpc.Core.Internal.Tests.AsyncC..."
```

The [`json.text`](https://github.com/grpc/grpc/blob/f14effb6e421601639ef9a293f3f9b55a51e87ed/src/csharp/tests.json) file contains not carriage return, and only the newline character (`\n`) as noted here - probably because it might have been generated on a Unix-type machine:

```Bash
$ grep $'\r' ./src/csharp/tests.json
$ grep $'\n' ./src/csharp/tests.json
{
  "Grpc.Core.Tests": [
    "Grpc.Core.Internal.Tests.AsyncCallServerTest",
    "Grpc.Core.Internal.Tests.AsyncCallTest",
    "Grpc.Core.Internal.Tests.ChannelArgsSafeHandleTest",
    "Grpc.Core.Internal.Tests.CompletionQueueEventTest",
    "Grpc.Core.Internal.Tests.CompletionQueueSafeHandleTest",
...
$
```

Hope it helps,
Paul

**Comment by apolcyn** [comment]
@jcanizales, @pgrosu, there is a fix for this on master, for which I also
have a copy that I needed to merge into v1.0.x too. It is here:
https://github.com/grpc/grpc/pull/7598

thanks,
Alex

**Comment by jcanizales** [comment]
Awesome. Next one is for @iamqizhao, a Go docker problem:

```
Cloning into 'src/google.golang.org/grpc'...
done.
Note: checking out '35896af9ad39c1fb1b1cd925fe3621be361e3d81'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b new_branch_name

/go/src/golang.org/x/oauth2/google/default.go:17:2: cannot find package "cloud.google.com/go/compute/metadata" in any of:
	/usr/local/go/src/cloud.google.com/go/compute/metadata (from $GOROOT)
	/go/src/cloud.google.com/go/compute/metadata (from $GOPATH)
+ EXITCODE=1
+ docker rm -f build_grpc_interop_go_e415a90d-f0e3-4ecd-b2ac-632062955733
build_grpc_interop_go_e415a90d-f0e3-4ecd-b2ac-632062955733
+ exit 1

FAILED: build_docker_go [ret=1, pid=15135]
```

**Comment by pgrosu** [comment]
@jcanizales  Awww, but the problem is too easy :)  How about a harder one for me?

**Comment by jcanizales** [comment]
@pgrosu I hope that's the last one! You're more than welcome to give me a fix, because I've never used docker and only once done a Go tutorial. Sounds like a Go package for Google Cloud Compute wasn't installed.

**Comment by pgrosu** [comment]
@jcanizales   I will gladly teach you Docker, but to fix this immediate issue, please add the following line just above line 45 of [build_interop.sh](https://github.com/grpc/grpc/blob/master/tools/dockerfile/interoptest/grpc_interop_go/build_interop.sh#L45) - yes, you are right it was missing [the package](https://godoc.org/google.golang.org/cloud/compute/metadata) to access GCE metadata, which is needed in [this file](https://github.com/golang/oauth2/blob/master/google/default.go#L17):

```
go get google.golang.org/cloud/compute/metadata
```

Basically the updated version should look something like this for those lines:

```
go get google.golang.org/cloud/compute/metadata
go get github.com/golang/protobuf/proto
go get golang.org/x/net/context
go get golang.org/x/net/trace
go get golang.org/x/oauth2
go get google.golang.org/cloud
```

If the `oath2` package requires more, then we might have it download the dependencies like this - but we can try that if the above does not work:

```
go get golang.org/x/oauth2/...
```

Let me know if that worked.

Thanks,
~p

**Comment by pgrosu** [comment]
@jcanizales  I have this strange suspicion that the directories are not being [mapped correctly](https://docs.docker.com/engine/tutorials/dockervolumes/) from the Jenkins host.  So to test this hypothesis, could you please update [line 95 of `build_interop_image.sh`](https://github.com/grpc/grpc/blob/master/tools/run_tests/dockerize/build_interop_image.sh#L95) to the following, and then rerun the build:

```Bash
bash -l -c 'go get google.golang.org/cloud/compute/metadata && bash /var/local/jenkins/grpc/tools/dockerfile/interoptest/$BASE_NAME/build_interop.sh' \
```

Sorry to bother you like this,
~p

**Comment by jcanizales** [comment]
Not bothered at all; I appreciate your help a lot.

**Comment by pgrosu** [comment]
@jcanizales  It is my fault, I just realized it is the same package with a slightly different path, so please replace it with the following specific one, and ignore my [bash test](https://github.com/grpc/grpc/pull/7475#issuecomment-237722881) above:

```
go get cloud.google.com/go/compute/metadata
```

The one I gave you before was the following - with `cloud` instead of `go` in the path - but it specifically is looking for the one above, even though their GoDoc descriptions seem to be same ([loc1](https://godoc.org/google.golang.org/cloud/compute/metadata) vs [loc2](https://godoc.org/cloud.google.com/go/compute/metadata)): 

```
google.golang.org/cloud/compute/metadata
```

So if you could try this small update and see if it fixes it, that would be great.

Sorry about that,
Paul

**Comment by pgrosu** [comment]
@jcanizales So it looks like the package issue was resolved #7653, so our fix https://github.com/grpc/grpc/pull/7475/commits/31871df58f2a99b21c7a85aebac73ed6d9bdaf9b can be removed and this PR rebased to match master.  Hopefully then resubmitting the other fixes in this PR to Jenkins for rebuilding would make it all work again :)

**Comment by pgrosu** [comment]
@jcanizales I think you might need to change the `build_interop.sh` file looks different than master - or rebase and resubmit the changes :)

From this PR:

https://github.com/jcanizales/grpc/blob/0db98cda0d81de2117501fd62020a7bbe6095eee/tools/dockerfile/interoptest/grpc_interop_go/build_interop.sh

On master:

https://github.com/grpc/grpc/blob/master/tools/dockerfile/interoptest/grpc_interop_go/build_interop.sh

There's also has a version by @menghanl that works at #7655 to try, if the above does not work.

**Comment by jcanizales** [comment]
Argh, I didn't see that PR #7653 was merged into master instead of this branch...

**Comment by pgrosu** [comment]
Yay the Interop tests pass :)

**Comment by pgrosu** [comment]
Have a great weekend Jorge!

**Comment by jcanizales** [comment]
You too, and thanks again for the help!
